Align diffusers CI tests with examples

dsocek opened the pull request:
# What does this PR do?

Comprehensive fixes and alignment between CI and examples README.md for diffusers

- 1st Commit re-organizes sections in README for better clarity and structure, and removes a number of over-represented samples (e.g. for controlnet and inpainting).  We also now have text-to-video integrated here rather than in standalone examples folder.   Please see improved README doc:
https://github.com/dsocek/optimum-habana/tree/align-diffusers-ci-tests-with-examples/examples/stable-diffusion

- 2nd Commit modifies tests/test_diffusers.py and closes a gap between README and ci.  A number of over-represented tests were removed and all slow tests were aligned to parameters used in README.  Also, some tests were merged into single test for efficiency improvement (e.g. separate tests which tested functionality and performance now merged to single test which runs pipeline once and then tests both)

yafshar commented on README.md:
```diff
@@ -271,7 +271,7 @@ The following model architectures, tasks and device distributions have been vali
 | LDM3D            |          | <li>Single card</li> | <li>[text-to-image generation](https://github.com/huggingface/optimum-habana/tree/main/examples/stable-diffusion)</li> |
 | FLUX.1           | <li>[fine-tuning](https://github.com/huggingface/optimum-habana/tree/main/examples/stable-diffusion/training#dreambooth-lora-fine-tuning-with-flux1-dev)</li> | <li>Single card</li> | <li>[text-to-image generation](https://github.com/huggingface/optimum-habana/tree/main/examples/stable-diffusion)</li> |
-| Text to Video    |          | <li>Single card</li> | <li>[text-to-video generation](https://github.com/huggingface/optimum-habana/tree/main/examples/text-to-video)</li> |
+| Text to Video    |          | <li>Single card</li> | <li>[text-to-video generation](https://github.com/huggingface/optimum-habana/tree/main/examples/stable-diffusion#text-to-video-generation)</li> |
```
@dsocek is there a reason you only added the section for `Text to Video` the rest only points to the `stable-diffusion` page

dsocek replied:
@yafshar I just moved text to video into stable diffusion folder and adjusted the README accordingly.  Ideally, agreed, we can be more specific in other links as well

yafshar commented on examples/stable-diffusion/README.md:
```diff
@@ -16,10 +16,10 @@ limitations under the License.
-This directory contains a script that showcases how to perform text-to-image generation using Stable Diffusion on Intel® Gaudi® AI Accelerators.
-
-Stable Diffusion was proposed in [Stable Diffusion Announcement](https://stability.ai/blog/stable-diffusion-announcement) by Patrick Esser and Robin Rombach and the Stability AI team.
+This directory contains sample scripts that showcase how to perform diffusion-based generative tasks on Intel® Gaudi® AI Accelerators.
```
```suggestion
This directory contains sample scripts demonstrating how to perform diffusion-based generative tasks on Intel® Gaudi® AI Accelerators.
```

yafshar commented on examples/stable-diffusion/README.md:
```diff
@@ -16,10 +16,10 @@ limitations under the License.
-Stable Diffusion was proposed in [Stable Diffusion Announcement](https://stability.ai/blog/stable-diffusion-announcement) by Patrick Esser and Robin Rombach and the Stability AI team.
+This directory contains sample scripts that showcase how to perform diffusion-based generative tasks on Intel® Gaudi® AI Accelerators.
 
+Stable Diffusion was proposed in [Stable Diffusion Announcement](https://stability.ai/blog/stable-diffusion-announcement) by Patrick Esser
+and Robin Rombach and the Stability AI team.
```
```suggestion
Stable Diffusion was introduced in [Stable Diffusion Announcement](https://stability.ai/blog/stable-diffusion-announcement) by Patrick Esser, Robin Rombach and the Stability AI team.
```

yafshar commented on examples/stable-diffusion/README.md:
```diff
@@ -28,11 +28,11 @@ First, you should install the requirements:
+### Stable Diffusion
 
-Here is how to generate images with one prompt:
+Here is how to generate images using Stable Diffusion 1.4 model with one prompt:
```
```suggestion
Here's how to generate images using the Stable Diffusion 1.4 model with a single prompt:
```

yafshar commented on examples/stable-diffusion/README.md:
```diff
@@ -48,13 +48,12 @@ python text_to_image_generation.py \
-### Multiple Prompts
-
-Here is how to generate images with several prompts:
+To generate images with several prompts, simply include 2 prompts in your input as follows:
```
```suggestion
To generate images with multiple prompts, simply include two prompts in your input as shown below:
```

yafshar commented on examples/stable-diffusion/README.md:
```diff
@@ -70,9 +69,8 @@ python text_to_image_generation.py \
-### Distributed inference with multiple HPUs
-
-Here is how to generate images with two prompts on two HPUs:
+Distributed inference with multiple HPUs is also supported. Here is an example which shows
+how to generate images with two prompts on two HPUs:
```
```suggestion
Distributed inference with multiple HPUs is also supported. Below is an example demonstrating how to generate images with two prompts on two HPUs:
```

yafshar commented on examples/stable-diffusion/README.md:
```diff
@@ -90,13 +88,14 @@ python ../gaudi_spawn.py \
-> HPU graphs are recommended when generating images by batches to get the fastest possible generations.
-> The first batch of images entails a performance penalty. All subsequent batches will be generated much faster.
-> You can enable this mode with `--use_hpu_graphs`.
+You can run other older Stable Diffusion models in a similar manner. For example, to perform image generation with
+Stable Diffusion 1.5, use the option: `--model_name_or_path stable-diffusion-v1-5/stable-diffusion-v1-5`. Examples
+showcasing Stable Diffusion 2 are given next.
```
```suggestion
You can run other older Stable Diffusion models in a similar manner. For example, to generate images with Stable Diffusion 1.5, use the option: `--model_name_or_path stable-diffusion-v1-5/stable-diffusion-v1-5`. Examples showcasing Stable Diffusion 2 are provided next.
```

yafshar commented on docs/source/index.mdx:
```diff
@@ -119,7 +119,7 @@ In the tables below, ✅ means single-card, multi-card and DeepSpeed have all be
 | LDM3D               |          | <div style="text-align:left"><li>Single card</li></div> | <li>[text-to-image generation](https://github.com/huggingface/optimum-habana/tree/main/examples/stable-diffusion)</li> |
 | FLUX.1              | <li>[fine-tuning](https://github.com/huggingface/optimum-habana/tree/main/examples/stable-diffusion/training#dreambooth-lora-fine-tuning-with-flux1-dev)</li> | <li>Single card</li> | <li>[text-to-image generation](https://github.com/huggingface/optimum-habana/tree/main/examples/stable-diffusion)</li> |
-| Text to Video       |          | <li>Single card</li> | <li>[text-to-video generation](https://github.com/huggingface/optimum-habana/tree/main/examples/text-to-video)</li> |
+| Text to Video       |          | <li>Single card</li> | <li>[text-to-video generation](https://github.com/huggingface/optimum-habana/tree/main/examples/stable-diffusion#text-to-video-generation)</li> |
```
@dsocek can you please fix the PATH here as your README

dsocek replied:
@yafshar I accidentally pushed a WIP to my branch by mistake, sorry about that!  Just fixed and pushed completed commit, so please re-check/re-review :)

yafshar commented on examples/stable-diffusion/README.md:
````diff
@@ -230,11 +212,11 @@ python text_to_image_generation.py \
 ```
 
-Here is how to generate SDXL images with two prompts on two HPUs:
+Distributed inferencing with Intel Gaudi accelerators is also supported for SDXL.  Here is an example of generating
+SDXL images in a distributed manner using two prompts on two HPUs:
````
```suggestion
SDXL also supports distributed inferencing with Intel Gaudi accelerators. Below is an example of generating SDXL images in a distributed manner using two prompts on two HPUs:
```

yafshar commented on examples/stable-diffusion/README.md:
```diff
@@ -187,30 +184,15 @@ python text_to_image_generation.py \
-SDXL combines a second text encoder (OpenCLIP ViT-bigG/14) with the original text encoder to significantly
-increase the number of parameters. Here is how to generate images with several prompts for both `prompt`
-and `prompt_2` (2nd text encoder), as well as their negative prompts:
+SDXL incorporates a second text encoder (OpenCLIP ViT-bigG/14) alongside the original Stable Diffusion text encoder,
+significantly increasing the number of parameters and enabling more detailed and descriptive prompts. Here is an
+example of how to generate images using multiple prompts for both `prompt` (primary text encoder) and `prompt_2`
+(secondary text encoder), along with their respective negative prompts:
```
```suggestion
SDXL integrates a second text encoder (OpenCLIP ViT-bigG/14), alongside the original Stable Diffusion text encoder. This addition significantly increases the number of parameters, enabling more detailed and descriptive prompts. Below is an example of how to generate images using multiple prompts for both `prompt` (primary text encoder) and `prompt_2` (secondary text encoder), along with their respective negative prompts:
```

yafshar commented on examples/stable-diffusion/README.md:
````diff
@@ -252,26 +234,14 @@ python ../gaudi_spawn.py \
-    --bf16 \
-    --optimize
-```
+The performance-optimized SDXL pipeline can be enabled using the `--optimize` option. This option leverages a more
+aggressively optimized modified attention mechanism for improved performance. Additionally, it supports running
````
```suggestion
The performance-optimized SDXL pipeline can be enabled using the `--optimize` option. This option utilizes a more aggressively optimized attention mechanism for enhanced performance. Additionally, it supports running
```

yafshar commented on examples/stable-diffusion/README.md:
```diff
@@ -252,26 +234,14 @@ python ../gaudi_spawn.py \
+inference in mixed FP8 precision.
 
-Here is how to generate SDXL images with optimized pipeline in fp8:
+Here is how to generate SDXL images with optimized pipeline in FP8 precision:
```
@dsocek this command fails!

```sh
    raise FileNotFoundError(f"Failed to load file {fname}")
FileNotFoundError: Failed to load file ./quantization/measure/fp8_hooks_maxabs.npz
```

Maybe we need to add a note or st like below

```txt
**Note:** Before executing the command below, please measure the model quantization statistics.
```

yafshar replied:
or maybe referring to `QUANT_CONFIG=./quantization/stable-diffusion-xl/measure_config.json`

dsocek replied:
Good catch.  I seem to forgot to update quantization `json` files with updated path after I moved them to sdxl subfolder.   Fixed now.

yafshar commented on examples/stable-diffusion/README.md:
```diff
@@ -286,13 +256,11 @@ QUANT_CONFIG=./quantization/quant_config.json python text_to_image_generation.py
+#### SDXL-Turbo
 
-SDXL-Turbo is a distilled version of SDXL 1.0, trained for real-time synthesis.
+Knowledge distillation technique can be used to train a distilled version of SDXL, where fewer
+inference steps are sufficient to generate high-quality images. SDXL-Turbo is a distilled
+version of Stable Diffusion XL 1.0, trained for real-time synthesis.
```
```suggestion
The knowledge distillation technique can be used to train a distilled version of SDXL, allowing for high-quality image generation with fewer inference steps. SDXL-Turbo is a distilled version of Stable Diffusion XL 1.0, optimized for real-time synthesis.
```

yafshar commented on examples/stable-diffusion/README.md:
```diff
@@ -449,34 +416,19 @@ python text_to_image_generation.py \
-It is a type of model for controlling StableDiffusion by conditioning the model with an additional input image.
-
-Here is how to generate images conditioned by canny edge model:
+Here is how to generate images conditioned by the canny edge model:
```
```suggestion
```

yafshar replied:
repeated a couple of lines below

yafshar commented on examples/stable-diffusion/README.md:
````diff
@@ -449,34 +416,19 @@ python text_to_image_generation.py \
-    --bf16
-```
+ControlNet was introduced in [Adding Conditional Control to Text-to-Image Diffusion Models](https://huggingface.co/papers/2302.05543)
+by Lvmin Zhang and Maneesh Agrawala. It enables conditioning the Stable Diffusion model with an additional input image, allowing
+for precise control over the composition of the generated images using various features such as edges, pose, depth, and more.
````
```suggestion
by Lvmin Zhang and Maneesh Agrawala. enables conditioning the Stable Diffusion model with an additional input image. This allows for precise control over the composition of generated images using various features such as edges, pose, depth, and more.
```

yafshar commented on examples/stable-diffusion/README.md:
```diff
@@ -488,11 +440,12 @@ python text_to_image_generation.py \
 
-Here is how to generate images conditioned by canny edge model and with two prompts on two HPUs:
+The ControlNet example can be run with multiple prompts by supplying more than one prompt in the input.
+In addition, it can also be run in a distributed manner. Here is how to generate images conditioned by
+Canny edge model and with two prompts on two HPUs:
```
```suggestion
Additionally, it supports distributed execution. Below is an example of generating images conditioned by the Canny edge model using two prompts on two HPUs:
```

yafshar commented on examples/stable-diffusion/README.md:
````diff
@@ -508,44 +461,9 @@ python ../gaudi_spawn.py \
-    --sdp_on_bf16 \
-    --bf16
-```
+These ControlNet examples will also do preprocessing of the input image and derive Canny edges from it. More generally,
+we can use `--control_preprocessing_type none` and supply preprocessed control image directly which allows running many 
+additional use cases.
````
```suggestion
These ControlNet examples will preprocess the input image to derive Canny edges. Alternatively, you can use `--control_preprocessing_type none` to supply a preprocessed control image directly, enabling many additional use cases.
```

yafshar commented on examples/stable-diffusion/README.md:
```diff
@@ -591,44 +509,107 @@ python text_to_image_generation.py \
+## Additional Stable Diffusion-based Inference Techniques
 
-### Single Prompt
+This section provides examples demonstrating additional inference techniques based on Stable Diffusion.
+For more details, please refer to
```
```suggestion
This section provides examples of additional inference techniques based on Stable Diffusion. For more details, please refer to
```

yafshar commented on examples/stable-diffusion/README.md:
```diff
@@ -591,44 +509,107 @@ python text_to_image_generation.py \
-Take instruct-pix2pix as an example.
+### Unconditional Image Generation
+
+Here is how to perform unconditional image generation on Intel Gaudi. For more details, please see
```
```suggestion
Here is how to perform unconditional image generation on Intel Gaudi. For more details,  please refer to the 
```

yafshar commented on examples/stable-diffusion/README.md:
```diff
@@ -591,44 +509,107 @@ python text_to_image_generation.py \
+
+### Controlling Brightness
+
+Here is how to run an example of controlling brightness. For more information, please refer to the
```
```suggestion
Here is an example of how to control brightness. For more information, please refer to the
```

yafshar commented on examples/stable-diffusion/README.md:
```diff
@@ -591,44 +509,107 @@ python text_to_image_generation.py \
+
+### Prompt Weighting
+
+Here is how to run an example of prompt weighting. For more information, please refer to the
```
```suggestion
Here is an example of how to run prompt weighting. For more information, please refer to the
```

yafshar commented on examples/stable-diffusion/README.md:
```diff
@@ -591,44 +509,107 @@ python text_to_image_generation.py \
+
+### Controlling Image Quality
+
+Here is how to run an example of improving image quality. For more details, please see
```
```suggestion
Here is an example of how to improve image quality. For more details, please see
```

yafshar commented on examples/stable-diffusion/README.md:
```diff
@@ -591,44 +509,107 @@ python text_to_image_generation.py \
+## Image-to-Image Generation
 
-### Multiple Prompts
+Images can also be generated from initial input images that guide the diffusion-based image generation process
```
```suggestion
Images can also be generated using initial input images to guide the diffusion-based image generation process.
```

yafshar commented on examples/stable-diffusion/README.md:
```diff
@@ -591,44 +509,107 @@ python text_to_image_generation.py \
-Here is how to generate images with several prompts and one image.
+### Stable Diffusion-based Image-to-Image
+
+Here is how to generate images with one prompt and one image using `timbrooks/instruct-pix2pix` model which is based on the Stable Diffusion:
```
```suggestion
Here is how to generate images using a single prompt and an input image with the `timbrooks/instruct-pix2pix` model, which is based on Stable Diffusion:
```

yafshar commented on examples/stable-diffusion/README.md:
```diff
@@ -642,13 +623,16 @@ python image_to_image_generation.py \
 ### Stable Diffusion XL Refiner
 
-Here is how to generate SDXL images with a single prompt and one image:
+Here is how to refine SDXL images with a single image and prompt:
```
```suggestion
Here is how to refine SDXL images using a single image and prompt:
```

yafshar commented on examples/stable-diffusion/README.md:
```diff
@@ -667,17 +651,17 @@ python image_to_image_generation.py \
+### FLUX.1 Image-to-Image
 
-Here is how to generate FLUX.1 images with a single prompt and one input image:
+Here is how to generate FLUX.1 image with one input image and a single prompt:
```
```suggestion
Here is how to generate a FLUX.1 image using a single input image and prompt:
```

yafshar commented on examples/stable-diffusion/README.md:
```diff
@@ -724,88 +708,20 @@ python depth_to_image_generation.py \
-Here is how to perform unconditional-image-generation on Gaudi/HPU.
-
-Original unconditional image generation pipeline is shared in here: [Unconditional Image Generation](https://huggingface.co/docs/diffusers/using-diffusers/unconditional_image_generation)
+This section showcases how to use the `GaudiTextToVideoSDPipeline` to run text-to-video generation tasks on HPUs.
+The pipeline uses a UNet3D structure and implements video generation through the iterative denoising process.
```
```suggestion
This section demonstrates how to use the `GaudiTextToVideoSDPipeline` for text-to-video generation tasks on HPUs.
The pipeline employs a UNet3D structure and generates videos through an iterative denoising process.
```

yafshar commented on examples/stable-diffusion/README.md:
```diff
@@ -724,88 +708,20 @@ python depth_to_image_generation.py \
-python unconditional_image_generation.py \
-    --model_name_or_path "google/ddpm-ema-celebahq-256" \
-    --batch_size 16 \
+python3 text_to_video_generation.py \
```
```suggestion
python text_to_video_generation.py \
```

yafshar commented:
@dsocek thanks for this PR. Great work. Have you run the CI for this PR? Please post the CI results and do the slow test with your modification. I am getting some failure testing.

yafshar commented:
The custom op issue has already been addressed in this pull request https://github.com/huggingface/optimum-habana/pull/1655

dsocek commented:
@yafshar all your changes are incorporated (I squashed many of these into 1 commit), and I also added commit which fixes the quant issue that you found

yafshar commented:
@dsocek thanks! Would you please post the test results? I am still seeing some issue. 

```sh
>>> RUN_SLOW=true GAUDI2_CI=1 python -m pytest tests/test_diffusers.py -s -v
```

yafshar commented on optimum/habana/diffusers/pipelines/pipeline_utils.py:
```diff
@@ -370,6 +368,16 @@ def from_pretrained(cls, pretrained_model_name_or_path: Optional[Union[str, os.P
         kwargs["bf16_full_eval"] = bf16_full_eval
 
+        # Need to load custom ops lists before instantiating htcore
+        if isinstance(kwargs["gaudi_config"], str):
```
```suggestion
        if isinstance(kwargs.get("gaudi_config"), str):
```

This way, if "gaudi_config" is not a key in kwargs, `kwargs.get("gaudi_config")` will return `None` by default, and `isinstance(None, str)` will return `False`.

yafshar replied:
This is an issue in one of the `test_textual_inversion` test

yafshar replied:
If "gaudi_config" is not a key in kwargs, we should also fix the else condition

dsocek replied:
@yafshar Thanks for catching this.   Actually I accidentally pushed this commit into this PR.  I have a separate PR for this fix so let's switch this discussion to there:  #1655.   I will remove it from this PR (doesn't belong here :)

yafshar replied:
@dsocek I suggested a change in https://github.com/huggingface/optimum-habana/pull/1655 adding it here, solved the issue 

```sh
>>> RUN_SLOW=true GAUDI2_CI=1 python -m pytest tests/test_diffusers.py -s -v -k  test_textual_inversion
================================== 2 passed, 151 deselected
```

dsocek replied:
@yafshar Thanks for great review.  I removed the 2 commits from this PR and updated #1655 with your suggested fix plus added my fix to handle cases where gaudi config is actually GaudiConfig type, like in [this example](https://github.com/huggingface/optimum-habana/blob/main/tests/test_diffusers.py#L690)

yafshar commented:
@dsocek thanks! Would you please post the test results? Everything else sounds great to me. We can ask regiss to check the PR.

dsocek commented:
@yafshar  Seems there were some issues on 1.19 (I originally tested only 1.18), related to SDP on BF16.  After PT 2.5 we now have SDP in full precision as opposed to bf16.  I added proper enablement for all diffusers and updated tests to use bf16 sdp.  Now all (slow and fast) diffusers tests pass on both 1.18 and 1.19.

```
export RUN_SLOW=true && \
export GAUDI2_CI=1 && \
make test_installs && \
python -m pytest tests/test_diffusers.py -v

Output:
tests/test_diffusers.py::GaudiPipelineUtilsTester::test_default PASSED                                                                                                                    [  0%]
tests/test_diffusers.py::GaudiPipelineUtilsTester::test_device PASSED                                                                                                                     [  1%]
tests/test_diffusers.py::GaudiPipelineUtilsTester::test_gaudi_config_raise_error_without_habana PASSED                                                                                    [  1%]
tests/test_diffusers.py::GaudiPipelineUtilsTester::test_gaudi_config_types PASSED                                                                                                         [  2%]
tests/test_diffusers.py::GaudiPipelineUtilsTester::test_save_pretrained PASSED                                                                                                            [  3%]
tests/test_diffusers.py::GaudiPipelineUtilsTester::test_use_hpu_graphs PASSED                                                                                                             [  3%]
tests/test_diffusers.py::GaudiPipelineUtilsTester::test_use_hpu_graphs_raise_error_without_habana PASSED                                                                                  [  4%]
tests/test_diffusers.py::GaudiStableDiffusionPipelineTester::test_no_generation_regression_ldm3d SKIPPED (test requires custom bf16 ops)                                                  [  5%]
tests/test_diffusers.py::GaudiStableDiffusionPipelineTester::test_no_generation_regression_upscale PASSED                                                                                 [  5%]
tests/test_diffusers.py::GaudiStableDiffusionPipelineTester::test_no_throughput_regression_autocast SKIPPED (test requires custom bf16 ops)                                               [  6%]
tests/test_diffusers.py::GaudiStableDiffusionPipelineTester::test_no_throughput_regression_bf16 PASSED                                                                                    [  7%]
tests/test_diffusers.py::GaudiStableDiffusionPipelineTester::test_stable_diffusion_batch_sizes PASSED                                                                                     [  7%]
tests/test_diffusers.py::GaudiStableDiffusionPipelineTester::test_stable_diffusion_bf16 PASSED                                                                                            [  8%]
tests/test_diffusers.py::GaudiStableDiffusionPipelineTester::test_stable_diffusion_ddim PASSED                                                                                            [  9%]
tests/test_diffusers.py::GaudiStableDiffusionPipelineTester::test_stable_diffusion_default PASSED                                                                                         [  9%]
tests/test_diffusers.py::GaudiStableDiffusionPipelineTester::test_stable_diffusion_hpu_graphs PASSED                                                                                      [ 10%]
tests/test_diffusers.py::GaudiStableDiffusionPipelineTester::test_stable_diffusion_no_safety_checker PASSED                                                                               [ 11%]
tests/test_diffusers.py::GaudiStableDiffusionPipelineTester::test_stable_diffusion_num_images_per_prompt PASSED                                                                           [ 11%]
tests/test_diffusers.py::GaudiStableDiffusionPipelineTester::test_stable_diffusion_output_types_0_pil PASSED                                                                              [ 12%]
tests/test_diffusers.py::GaudiStableDiffusionPipelineTester::test_stable_diffusion_output_types_1_np PASSED                                                                               [ 13%]
tests/test_diffusers.py::GaudiStableDiffusionPipelineTester::test_stable_diffusion_output_types_2_latent PASSED                                                                           [ 13%]
tests/test_diffusers.py::GaudiStableDiffusionPipelineTester::test_sd_textual_inversion SKIPPED (system does not have 8 cards)                                                             [ 14%]
tests/test_diffusers.py::GaudiStableDiffusionXLPipelineTester::test_stable_diffusion_xl_batch_sizes PASSED                                                                                [ 15%]
tests/test_diffusers.py::GaudiStableDiffusionXLPipelineTester::test_stable_diffusion_xl_bf16 PASSED                                                                                       [ 15%]
tests/test_diffusers.py::GaudiStableDiffusionXLPipelineTester::test_stable_diffusion_xl_default PASSED                                                                                    [ 16%]
tests/test_diffusers.py::GaudiStableDiffusionXLPipelineTester::test_stable_diffusion_xl_euler PASSED                                                                                      [ 16%]
tests/test_diffusers.py::GaudiStableDiffusionXLPipelineTester::test_stable_diffusion_xl_euler_ancestral PASSED                                                                            [ 17%]
tests/test_diffusers.py::GaudiStableDiffusionXLPipelineTester::test_stable_diffusion_xl_generation_throughput PASSED                                                                      [ 18%]
tests/test_diffusers.py::GaudiStableDiffusionXLPipelineTester::test_stable_diffusion_xl_hpu_graphs PASSED                                                                                 [ 18%]
tests/test_diffusers.py::GaudiStableDiffusionXLPipelineTester::test_stable_diffusion_xl_num_images_per_prompt PASSED                                                                      [ 19%]
tests/test_diffusers.py::GaudiStableDiffusionXLPipelineTester::test_stable_diffusion_xl_output_types_0_pil PASSED                                                                         [ 20%]
tests/test_diffusers.py::GaudiStableDiffusionXLPipelineTester::test_stable_diffusion_xl_output_types_1_np PASSED                                                                          [ 20%]
tests/test_diffusers.py::GaudiStableDiffusionXLPipelineTester::test_stable_diffusion_xl_output_types_2_latent PASSED                                                                      [ 21%]
tests/test_diffusers.py::GaudiStableDiffusionXLPipelineTester::test_stable_diffusion_xl_turbo_euler_ancestral PASSED                                                                      [ 22%]
tests/test_diffusers.py::GaudiStableDiffusionXLPipelineTester::test_sdxl_textual_inversion PASSED                                                                                         [ 22%]
tests/test_diffusers.py::GaudiStableDiffusion3PipelineTester::test_fused_qkv_projections PASSED                                                                                           [ 23%]
tests/test_diffusers.py::GaudiStableDiffusion3PipelineTester::test_stable_diffusion_3_different_negative_prompts PASSED                                                                   [ 24%]
tests/test_diffusers.py::GaudiStableDiffusion3PipelineTester::test_stable_diffusion_3_different_prompts PASSED                                                                            [ 24%]
tests/test_diffusers.py::GaudiStableDiffusion3PipelineTester::test_stable_diffusion_3_prompt_embeds PASSED                                                                                [ 25%]
tests/test_diffusers.py::GaudiStableDiffusionControlNetPipelineTester::test_stable_diffusion_controlnet_batch_sizes PASSED                                                                [ 26%]
tests/test_diffusers.py::GaudiStableDiffusionControlNetPipelineTester::test_stable_diffusion_controlnet_bf16 PASSED                                                                       [ 26%]
tests/test_diffusers.py::GaudiStableDiffusionControlNetPipelineTester::test_stable_diffusion_controlnet_default PASSED                                                                    [ 27%]
tests/test_diffusers.py::GaudiStableDiffusionControlNetPipelineTester::test_stable_diffusion_controlnet_hpu_graphs PASSED                                                                 [ 28%]
tests/test_diffusers.py::GaudiStableDiffusionControlNetPipelineTester::test_stable_diffusion_controlnet_num_images_per_prompt PASSED                                                      [ 28%]
tests/test_diffusers.py::GaudiStableDiffusionMultiControlNetPipelineTester::test_stable_diffusion_multicontrolnet_batch_sizes PASSED                                                      [ 29%]
tests/test_diffusers.py::GaudiStableDiffusionMultiControlNetPipelineTester::test_stable_diffusion_multicontrolnet_bf16 PASSED                                                             [ 30%]
tests/test_diffusers.py::GaudiStableDiffusionMultiControlNetPipelineTester::test_stable_diffusion_multicontrolnet_default PASSED                                                          [ 30%]
tests/test_diffusers.py::GaudiStableDiffusionMultiControlNetPipelineTester::test_stable_diffusion_multicontrolnet_hpu_graphs PASSED                                                       [ 31%]
tests/test_diffusers.py::GaudiStableDiffusionMultiControlNetPipelineTester::test_stable_diffusion_multicontrolnet_num_images_per_prompt PASSED                                            [ 32%]
tests/test_diffusers.py::GaudiStableDiffusionDepth2ImgPipelineTester::test_depth2img_pipeline PASSED                                                                                      [ 32%]
tests/test_diffusers.py::GaudiStableDiffusionDepth2ImgPipelineTester::test_depth2img_pipeline_batch PASSED                                                                                [ 33%]
tests/test_diffusers.py::GaudiStableDiffusionDepth2ImgPipelineTester::test_depth2img_pipeline_bf16 PASSED                                                                                 [ 33%]
tests/test_diffusers.py::GaudiStableDiffusionDepth2ImgPipelineTester::test_depth2img_pipeline_default PASSED                                                                              [ 34%]
tests/test_diffusers.py::GaudiStableDiffusionDepth2ImgPipelineTester::test_depth2img_pipeline_hpu_graphs PASSED                                                                           [ 35%]
tests/test_diffusers.py::TrainTextToImage::test_train_text_to_image_script PASSED                                                                                                         [ 35%]
tests/test_diffusers.py::TrainTextToImage::test_train_text_to_image_sdxl PASSED                                                                                                           [ 36%]
tests/test_diffusers.py::TrainControlNet::test_train_controlnet SKIPPED (system does not have 8 cards)                                                                                    [ 37%]
tests/test_diffusers.py::TrainControlNet::test_script_train_controlnet PASSED                                                                                                             [ 37%]
tests/test_diffusers.py::DreamBooth::test_dreambooth_full PASSED                                                                                                                          [ 38%]
tests/test_diffusers.py::DreamBooth::test_dreambooth_full_with_text_encoder PASSED                                                                                                        [ 39%]
tests/test_diffusers.py::DreamBooth::test_dreambooth_loha PASSED                                                                                                                          [ 39%]
tests/test_diffusers.py::DreamBooth::test_dreambooth_loha_with_text_encoder PASSED                                                                                                        [ 40%]
tests/test_diffusers.py::DreamBooth::test_dreambooth_lokr PASSED                                                                                                                          [ 41%]
tests/test_diffusers.py::DreamBooth::test_dreambooth_lokr_with_text_encoder PASSED                                                                                                        [ 41%]
tests/test_diffusers.py::DreamBooth::test_dreambooth_lora PASSED                                                                                                                          [ 42%]
tests/test_diffusers.py::DreamBooth::test_dreambooth_lora_with_text_encoder PASSED                                                                                                        [ 43%]
tests/test_diffusers.py::DreamBooth::test_dreambooth_oft PASSED                                                                                                                           [ 43%]
tests/test_diffusers.py::DreamBooth::test_dreambooth_oft_with_text_encoder PASSED                                                                                                         [ 44%]
tests/test_diffusers.py::DreamBoothLoRASDXL::test_dreambooth_lora_sdxl PASSED                                                                                                             [ 45%]
tests/test_diffusers.py::DreamBoothLoRASDXL::test_dreambooth_lora_sdxl_with_text_encoder PASSED                                                                                           [ 45%]
tests/test_diffusers.py::GaudiStableVideoDiffusionPipelineTester::test_stable_video_diffusion_no_throughput_regression_bf16 PASSED                                                        [ 46%]
tests/test_diffusers.py::GaudiStableVideoDiffusionPipelineTester::test_stable_video_diffusion_single_video PASSED                                                                         [ 47%]
tests/test_diffusers.py::GaudiStableVideoDiffusionControlNetPipelineTester::test_stable_video_diffusion_single_video PASSED                                                               [ 47%]
tests/test_diffusers.py::GaudiStableDiffusionInstructPix2PixPipelineTests::test_stable_diffusion_pix2pix_default_case PASSED                                                              [ 48%]
tests/test_diffusers.py::GaudiStableDiffusionInstructPix2PixPipelineTests::test_stable_diffusion_pix2pix_euler PASSED                                                                     [ 49%]
tests/test_diffusers.py::GaudiStableDiffusionInstructPix2PixPipelineTests::test_stable_diffusion_pix2pix_multiple_init_images PASSED                                                      [ 49%]
tests/test_diffusers.py::GaudiStableDiffusionInstructPix2PixPipelineTests::test_stable_diffusion_pix2pix_negative_prompt PASSED                                                           [ 50%]
tests/test_diffusers.py::GaudiStableDiffusionImg2ImgPipelineTests::test_stable_diffusion_img2img_default_case PASSED                                                                      [ 50%]
tests/test_diffusers.py::GaudiStableDiffusionImg2ImgPipelineTests::test_stable_diffusion_img2img_multiple_init_images PASSED                                                              [ 51%]
tests/test_diffusers.py::GaudiStableDiffusionImg2ImgPipelineTests::test_stable_diffusion_img2img_negative_prompt PASSED                                                                   [ 52%]
tests/test_diffusers.py::GaudiStableDiffusionImageVariationPipelineTests::test_stable_diffusion_img_variation_default_case PASSED                                                         [ 52%]
tests/test_diffusers.py::GaudiStableDiffusionImageVariationPipelineTests::test_stable_diffusion_img_variation_multiple_images PASSED                                                      [ 53%]
tests/test_diffusers.py::GaudiStableDiffusionXLImg2ImgPipelineTests::test_components_function PASSED                                                                                      [ 54%]
tests/test_diffusers.py::GaudiStableDiffusionXLImg2ImgPipelineTests::test_stable_diffusion_xl_img2img_euler PASSED                                                                        [ 54%]
tests/test_diffusers.py::GaudiTextToVideoSDPipelineTester::test_stable_video_diffusion_no_latency_regression_bf16 PASSED                                                                  [ 55%]
tests/test_diffusers.py::GaudiTextToVideoSDPipelineTester::test_text_to_video_default_case PASSED                                                                                         [ 56%]
tests/test_diffusers.py::StableDiffusionInpaintPipelineTests::test_attention_slicing_forward_pass PASSED                                                                                  [ 56%]
tests/test_diffusers.py::StableDiffusionInpaintPipelineTests::test_callback_cfg PASSED                                                                                                    [ 57%]
tests/test_diffusers.py::StableDiffusionInpaintPipelineTests::test_callback_inputs PASSED                                                                                                 [ 58%]
tests/test_diffusers.py::StableDiffusionInpaintPipelineTests::test_cfg PASSED                                                                                                             [ 58%]
tests/test_diffusers.py::StableDiffusionInpaintPipelineTests::test_components_function PASSED                                                                                             [ 59%]
tests/test_diffusers.py::StableDiffusionInpaintPipelineTests::test_dict_tuple_outputs_equivalent PASSED                                                                                   [ 60%]
tests/test_diffusers.py::StableDiffusionInpaintPipelineTests::test_inference_batch_consistent PASSED                                                                                      [ 61%]
tests/test_diffusers.py::StableDiffusionInpaintPipelineTests::test_inference_batch_single_identical PASSED                                                                                [ 62%]
tests/test_diffusers.py::StableDiffusionInpaintPipelineTests::test_karras_schedulers_shape PASSED                                                                                         [ 62%]
tests/test_diffusers.py::StableDiffusionInpaintPipelineTests::test_latents_input PASSED                                                                                                   [ 63%]
tests/test_diffusers.py::StableDiffusionInpaintPipelineTests::test_num_images_per_prompt PASSED                                                                                           [ 64%]
tests/test_diffusers.py::StableDiffusionInpaintPipelineTests::test_pipeline_call_signature PASSED                                                                                         [ 65%]
tests/test_diffusers.py::StableDiffusionInpaintPipelineTests::test_progress_bar PASSED                                                                                                    [ 66%]
tests/test_diffusers.py::StableDiffusionInpaintPipelineTests::test_pt_np_pil_inputs_equivalent PASSED                                                                                     [ 66%]
tests/test_diffusers.py::StableDiffusionInpaintPipelineTests::test_pt_np_pil_outputs_equivalent PASSED                                                                                    [ 67%]
tests/test_diffusers.py::StableDiffusionInpaintPipelineTests::test_stable_diffusion_inpaint PASSED                                                                                        [ 69%]
tests/test_diffusers.py::StableDiffusionInpaintPipelineTests::test_stable_diffusion_inpaint_no_throughput_regression PASSED                                                               [ 69%]
tests/test_diffusers.py::StableDiffusionInpaintPipelineTests::test_to_dtype PASSED                                                                                                        [ 71%]
tests/test_diffusers.py::StableDiffusionXLInpaintPipelineTests::test_attention_slicing_forward_pass PASSED                                                                                [ 72%]
tests/test_diffusers.py::StableDiffusionXLInpaintPipelineTests::test_callback_cfg PASSED                                                                                                  [ 73%]
tests/test_diffusers.py::StableDiffusionXLInpaintPipelineTests::test_callback_inputs PASSED                                                                                               [ 73%]
tests/test_diffusers.py::StableDiffusionXLInpaintPipelineTests::test_cfg PASSED                                                                                                           [ 74%]
tests/test_diffusers.py::StableDiffusionXLInpaintPipelineTests::test_components_function PASSED                                                                                           [ 75%]
tests/test_diffusers.py::StableDiffusionXLInpaintPipelineTests::test_dict_tuple_outputs_equivalent PASSED                                                                                 [ 75%]
tests/test_diffusers.py::StableDiffusionXLInpaintPipelineTests::test_inference_batch_consistent PASSED                                                                                    [ 77%]
tests/test_diffusers.py::StableDiffusionXLInpaintPipelineTests::test_inference_batch_single_identical PASSED                                                                              [ 77%]
tests/test_diffusers.py::StableDiffusionXLInpaintPipelineTests::test_latents_input PASSED                                                                                                 [ 78%]
tests/test_diffusers.py::StableDiffusionXLInpaintPipelineTests::test_num_images_per_prompt PASSED                                                                                         [ 79%]
tests/test_diffusers.py::StableDiffusionXLInpaintPipelineTests::test_pipeline_call_signature PASSED                                                                                       [ 80%]
tests/test_diffusers.py::StableDiffusionXLInpaintPipelineTests::test_pipeline_interrupt PASSED                                                                                            [ 81%]
tests/test_diffusers.py::StableDiffusionXLInpaintPipelineTests::test_progress_bar PASSED                                                                                                  [ 81%]
tests/test_diffusers.py::StableDiffusionXLInpaintPipelineTests::test_pt_np_pil_inputs_equivalent PASSED                                                                                   [ 82%]
tests/test_diffusers.py::StableDiffusionXLInpaintPipelineTests::test_pt_np_pil_outputs_equivalent PASSED                                                                                  [ 83%]
tests/test_diffusers.py::StableDiffusionXLInpaintPipelineTests::test_stable_diffusion_two_xl_mixture_of_denoiser_fast PASSED                                                              [ 84%]
tests/test_diffusers.py::StableDiffusionXLInpaintPipelineTests::test_stable_diffusion_xl_img2img_negative_conditions PASSED                                                               [ 85%]
tests/test_diffusers.py::StableDiffusionXLInpaintPipelineTests::test_stable_diffusion_xl_inpaint_2_images PASSED                                                                          [ 86%]
tests/test_diffusers.py::StableDiffusionXLInpaintPipelineTests::test_stable_diffusion_xl_inpaint_euler PASSED                                                                             [ 86%]
tests/test_diffusers.py::StableDiffusionXLInpaintPipelineTests::test_stable_diffusion_xl_inpaint_euler_lcm PASSED                                                                         [ 87%]
tests/test_diffusers.py::StableDiffusionXLInpaintPipelineTests::test_stable_diffusion_xl_inpaint_euler_lcm_custom_timesteps PASSED                                                        [ 88%]
tests/test_diffusers.py::StableDiffusionXLInpaintPipelineTests::test_stable_diffusion_xl_inpaint_mask_latents PASSED                                                                      [ 88%]
tests/test_diffusers.py::StableDiffusionXLInpaintPipelineTests::test_stable_diffusion_xl_inpaint_negative_prompt_embeds PASSED                                                            [ 89%]
tests/test_diffusers.py::StableDiffusionXLInpaintPipelineTests::test_stable_diffusion_xl_inpaint_no_throughput_regression PASSED                                                          [ 90%]
tests/test_diffusers.py::StableDiffusionXLInpaintPipelineTests::test_stable_diffusion_xl_multi_prompts PASSED                                                                             [ 90%]
tests/test_diffusers.py::StableDiffusionXLInpaintPipelineTests::test_stable_diffusion_xl_refiner PASSED                                                                                   [ 91%]
tests/test_diffusers.py::StableDiffusionXLInpaintPipelineTests::test_to_dtype PASSED                                                                                                      [ 92%]
tests/test_diffusers.py::GaudiDDPMPipelineTester::test_ddpmpipline_batch_sizes PASSED                                                                                                     [ 94%]
tests/test_diffusers.py::GaudiDDPMPipelineTester::test_ddpmpipline_bf16 PASSED                                                                                                            [ 94%]
tests/test_diffusers.py::GaudiDDPMPipelineTester::test_ddpmpipline_default PASSED                                                                                                         [ 95%]
tests/test_diffusers.py::GaudiDDPMPipelineTester::test_ddpmpipline_hpu_graphs PASSED                                                                                                      [ 96%]
tests/test_diffusers.py::GaudiDDPMPipelineTester::test_no_throughput_regression_bf16 PASSED                                                                                               [ 96%]
tests/test_diffusers.py::GaudiFluxPipelineTester::test_flux_different_prompts PASSED                                                                                                      [ 97%]
tests/test_diffusers.py::GaudiFluxPipelineTester::test_flux_inference PASSED                                                                                                              [ 98%]
tests/test_diffusers.py::GaudiFluxPipelineTester::test_flux_prompt_embeds PASSED                                                                                                          [ 98%]
tests/test_diffusers.py::GaudiFluxImg2ImgPipelineTester::test_flux_different_prompts PASSED                                                                                               [ 99%]
tests/test_diffusers.py::GaudiFluxImg2ImgPipelineTester::test_flux_prompt_embeds PASSED                                                                                                   [100%]
```

Multi-card tests:
```
python -m pytest tests/test_diffusers.py -v -k "test_sd_textual_inversion"
Output:
tests/test_diffusers.py::GaudiStableDiffusionPipelineTester::test_sd_textual_inversion PASSED                                                                                             [100%]

python -m pytest tests/test_diffusers.py -v -k "test_sd_textual_inversion"
Output:
tests/test_diffusers.py::TrainControlNet::test_train_controlnet PASSED                                                                                                                    [100%]
```

yafshar commented:
@dsocek I am still getting some random fails from ''textual_inversion" test, but I think that overall everything is great. We should merge this PR and fix anything else later.

```sh
[rank0]:   File "/usr/local/lib/python3.10/dist-packages/habana_frameworks/torch/core/weight_sharing.py", line 75, in __torch_function__
[rank0]:     return super().__torch_function__(func, types, new_args, kwargs)
[rank0]:   File "/usr/local/lib/python3.10/dist-packages/habana_frameworks/torch/hpu/__init__.py", line 81, in _lazy_init
[rank0]:     _hpu_C.init()
[rank0]: RuntimeError: synStatus=8 [Device not found] Device acquire failed.
```

yafshar reviewed (approved):
LGTM!

Hi @regisss, this PR is ready for your final review. Could you please take a look?

dsocek commented:
> @dsocek I am still getting some random fails from ''textual_inversion" test, but I think that overall everything is great. We should merge this PR and fix anything else later.
> 
> ```shell
> [rank0]:   File "/usr/local/lib/python3.10/dist-packages/habana_frameworks/torch/core/weight_sharing.py", line 75, in __torch_function__
> [rank0]:     return super().__torch_function__(func, types, new_args, kwargs)
> [rank0]:   File "/usr/local/lib/python3.10/dist-packages/habana_frameworks/torch/hpu/__init__.py", line 81, in _lazy_init
> [rank0]:     _hpu_C.init()
> [rank0]: RuntimeError: synStatus=8 [Device not found] Device acquire failed.
> ```

@yafshar actually this is some issue which I noticed.  The CI will fail if you run multi-card tests with other tests.  However, if you run it alone, it will pass.   Should be investigated and fixed independently of this PR

regisss commented:
> > @dsocek I am still getting some random fails from ''textual_inversion" test, but I think that overall everything is great. We should merge this PR and fix anything else later.
> > ```shell
> > [rank0]:   File "/usr/local/lib/python3.10/dist-packages/habana_frameworks/torch/core/weight_sharing.py", line 75, in __torch_function__
> > [rank0]:     return super().__torch_function__(func, types, new_args, kwargs)
> > [rank0]:   File "/usr/local/lib/python3.10/dist-packages/habana_frameworks/torch/hpu/__init__.py", line 81, in _lazy_init
> > [rank0]:     _hpu_C.init()
> > [rank0]: RuntimeError: synStatus=8 [Device not found] Device acquire failed.
> > ```
> 
> @yafshar actually this is some issue which I noticed. The CI will fail if you run multi-card tests with other tests. However, if you run it alone, it will pass. Should be investigated and fixed independently of this PR

Yes, this is something I have noticed too and that's actually why the Diffusers slow tests are not run with one command but several specific ones: https://github.com/huggingface/optimum-habana/blob/3c251d0480d916bcb08d0bda092edf03847ca11b/Makefile#L101

regisss commented on tests/test_diffusers.py:
```diff
@@ -135,32 +131,30 @@
-    TEXT_TO_VIDEO_SYNTHESIS_BF16_BASELINE = 70
-    DETERMINISTIC_IMAGE_GENERATION_THROUGHPUT = 0.946
-    THROUGHPUT_UNCONDITIONAL_IMAGE_BASELINE_BF16 = 0.15186785472532677
-    DEPTH2IMG_GENERATION_LATENCY_BASELINE_BF16 = 36.06376791000366
```
Not opposed to it, but why removing this one or `DETERMINISTIC_IMAGE_GENERATION_THROUGHPUT` ?

dsocek replied:
We removed some outdated and less main-stream samples/tests, this is one of those.   Trying to minimize number of slow tests and keep basic pipeline slow tests.

regisss replied:
Sounds good!

regisss reviewed (approved):
LGTM!
